### PR TITLE
Issues/3786 post list sorting

### PR DESCRIPTION
**Pull request by aerych**
Refs #3786
* Requires #3805 *

This PR builds on work done in #3805.  The relevant changes begin in dd7fa34. 

The goal is to add two new helper or "meta" properties to help with sorting posts and pages returned from an NSFetchRequest. The new properties are automatically updated in response to changes to a posts date created or its remote status. 

To ensure any existing posts get the correct values set, this PR includes a core data revision, mapping model, and migration policy for posts and pages. 

### TESTING
Setup:  Prepare a local only post in the simulator *BEFORE* switching to this branch. For convenience, be sure to use a blog with very view posts  Follow the steps outlined in this video to create a local-only post https://cloudup.com/iOZ-BQXHqyx 

Test 1: 
- Launch the simulator. Confirm there are no errors launching the app and you do not have to sign in again. 
- View the post list.  The local only post should be first in the list. 

Test 2: 
- Use a [link conditioner](http://nshipster.com/network-link-conditioner/) to reduce your network speed.  A slow connection will help you verify this is working. 
- Edit an existing post. Its easier if its near the top of the list, but not THE first one in the list. 
- Set it to publish immediately and save your changes. 
- Watch closely and confirm that it is the first post in the list when publishing. 

Needs Review: @koke @astralbodies

### CODE REVIEW

**Comment by aerych** [comment]
@astralbodies Would really appreciate your eyes on the migration policy to make sure its a good approach.

File: WordPress/Classes/Models/AbstractPost.m [hunk 1]
```diff
@@ -22,7 +26,28 @@ - (void)awakeFromFetch
...
+
+- (void)setDate_created_gmt:(NSDate *)date_created_gmt
+{
+NSLog(@"%@", NSStringFromSelector(_cmd));
+    self.metaPublishImmediately = (date_created_gmt == nil);
+
+    NSString *key = @"date_created_gmt";
+    [self willChangeValueForKey:key];
+    [self setPrimitiveValue:date_created_gmt forKey:key];
+    [self didChangeValueForKey:key];
```
Comment (koke) [inline WordPress/Classes/Models/AbstractPost.m]: Since you're supporting KVO, I recently learned that you should register dependent keys with `keyPathsForValuesAffecting*`

https://developer.apple.com/library/ios/documentation/Cocoa/Conceptual/KeyValueObserving/Articles/KVODependentKeys.html#//apple_ref/doc/uid/20002179-BAJEAIEE

File: WordPress/Classes/Models/AbstractPost.m [hunk 2]
```diff
@@ -22,7 +26,28 @@ - (void)awakeFromFetch
         // when wrong saving -- the app crashed for instance. So change our remote status to failed.
         [self setPrimitiveValue:@(AbstractPostRemoteStatusFailed) forKey:@"remoteStatusNumber"];
     }
+}
+
+- (void)setRemoteStatusNumber:(NSNumber *)remoteStatusNumber
+{
+NSLog(@"%@", NSStringFromSelector(_cmd));
```
Comment (koke) [inline WordPress/Classes/Models/AbstractPost.m]: No `NSLog` in production :)

File: WordPress/Classes/Models/AbstractPost.m [hunk 3]
```diff
@@ -22,7 +26,28 @@ - (void)awakeFromFetch
...
 
+    NSString *key = @"remoteStatusNumber";
+    [self willChangeValueForKey:key];
+    [self setPrimitiveValue:remoteStatusNumber forKey:key];
+    [self didChangeValueForKey:key];
+}
+
+- (void)setDate_created_gmt:(NSDate *)date_created_gmt
+{
+NSLog(@"%@", NSStringFromSelector(_cmd));
```
Comment (koke) [inline WordPress/Classes/Models/AbstractPost.m]: No `NSLog` in production :)

File: WordPress/Classes/Models/AbstractPost.m (hunk 1 above)
Comment (koke) [inline WordPress/Classes/Models/AbstractPost.m]: And I think the `self.meta* = ...` should be between the `willChange` and the `didChange`

File: WordPress/Classes/Models/AbstractPost.m (hunk 2 above)
Comment (aerych) [inline WordPress/Classes/Models/AbstractPost.m]: oopsie!

**Comment by koke** [comment]
I might be missing something, but the `metaPublishImmediately` property looks redundant. In my head, local posts is something that's there so you don't lose content if the app gets killed, so it should only ever be one or two of those. Having a second sort descriptor feels like adding more complexity than it's worth.

File: WordPress/Classes/Models/AbstractPost.m (hunk 1 above)
Comment (aerych) [inline WordPress/Classes/Models/AbstractPost.m]: > keyPathsForValuesAffecting

Nice find.

**Comment by astralbodies** [comment]
@aerych This looks okay. I did add a very BASIC unit test to cover migrations from 19 through 31. This does not ensure the data is migrated properly, however. Can you update the unit test to be something similar to `testMigrate27to28preservesCategories` in `CoreDataMigrationTests` to make sure your intended behavior happens?

**Comment by aerych** [comment]
> I did add a very BASIC unit test to cover migrations from 19 through 31 

Awesome, thanks for that! I'll have a look at the test you mentioned.

**Comment by aerych** [comment]
> I might be missing something, but the metaPublishImmediately property looks redundant

Its intended to cover the edge case when a non-local post (e.g. a draft that already has a remote) is edited to publish immediately, and encounters an error when publishing leaving it in a failed state. In this case the create date would still be nil, but the meta property would be flagged to indicate the correct sort order.

**Comment by aerych** [comment]
@astralbodies I've updated the test. Ready for another round.

File: WordPress/Classes/Utility/Migrations/PostToPost30To31.m [hunk 4]
```diff
@@ -0,0 +1,54 @@
+#import "PostToPost30To31.h"
+#import "AbstractPost.h"
```
Comment (astralbodies) [inline WordPress/Classes/Utility/Migrations/PostToPost30To31.m]: Make sure to not use any model objects in the migration as they are only valid on the most recent model version. I'd remove this import even though its not used.

Comment (aerych) [inline WordPress/Classes/Utility/Migrations/PostToPost30To31.m]: :+1:

**Comment by astralbodies** [comment]
:shipit:

**Comment by aerych** [comment]
Thanks guys!!
